### ISSUE TITLE
Periodic upgrade jobs broken in testgrid

### ISSUE THREAD

**Issue by mboersma** [kubernetes-sigs/cluster-api-provider-azure#2693]
/kind bug

**What steps did you take and what happened:**

All the capz-periodic-upgrade-main-* jobs on CAPZ's [testgrid dashboard](https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure) are broken. No workload clusters are being provisioned. Testgrid points the finger at [these changes](https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/89b354a2c...4c28715a5?).

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by mboersma** [comment]
Our theory is that we're getting a "name longer than 63 chars" error trying to create a resource from the string created [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/helpers.go#L605), which is new code added in #2661.

cc: @CecileRobertMichon

**Comment by jackfrancis** [comment]
Is this a 63 char limit a CRD limitation?

**Comment by mboersma** [comment]
Not sure exactly where it's being validated @jackfrancis, but here's an error from the capz-controller-manager log for one [failed job](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-provider-azure-e2e-workload-upgrade-1-22-1-23-main/1576388748772380672):

```
E1002 02:00:15.353339       1 controller.go:326]  "msg"="Reconciler error" "error"="AzureClusterIdentity.infrastructure.cluster.x-k8s.io \"cluster-identity-sp\" is invalid: metadata.finalizers: Invalid value: \"azurecluster.infrastructure.cluster.x-k8s.io/k8s-upgrade-and-conformance-2xntab-k8s-upgrade-and-conformance-9k7lig\": name part must be no more than 63 characters" "azureCluster"={"name":"k8s-upgrade-and-conformance-9k7lig","namespace":"k8s-upgrade-and-conformance-2xntab"} "controller"="azurecluster" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureCluster" "name"="k8s-upgrade-and-conformance-9k7lig" "namespace"="k8s-upgrade-and-conformance-2xntab" "reconcileID"="e3a06c57-8d50-4182-9d9f-dbe89878272d" 
```

**Comment by jackfrancis** [comment]
Indeed this 8 bit requirement is all over kubernetes source, we should be able to fix w/ a less onerous name in our tests, or a more concise concatenation strategy.

Good find @mboersma!

**Comment by CecileRobertMichon** [comment]
/assign

The root cause here is that finalizer string length that I added for AzureClusterIdentity has variable length: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/helpers.go#L605

It's currently in the format:

```
prefix/clusterNamespace-clusterName
```

where prefix is the api group of the owning cluster (either `azurecluster.infrastructure.cluster.x-k8s.io` or `azuremanagedcontrolplane.infrastructure.cluster.x-k8s.io`) and then cluster namespace/name is the name/namespace of the owning cluster.

The main reason I did this was to avoid a race condition where a cluster identity is used by more than one cluster and they are all being deleted at the same time. If the two clusters share the same finalizer, it would cause the finalizer to be deleted as soon as the first cluster finishes deleting, which would cause the second cluster to not be able to finish deleting. 

There are a few solutions to fix/mitigate this, I'm not particularly sold on either so open to suggestions:

1) Remove the prefix altogether: length would still be variable but a lot shorter to start so less chances of running into the max, no differentiation on resource type, won't follow typical finalizer naming conventions
2) Remove cluster name and/or namespace: Finalizer key would be fixed, but introduces a race condition, I don't think we should do this.
3) If name is longer than 63 chars, create a hash or trim the key. Would make sure we don't ever run into the limit. Less user friendly.
